### PR TITLE
WIP Add temporary databases  https://github.com/ClickHouse/ClickHouse/issues/31497

### PR DESCRIPTION
**Pull request by Lirikl**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
add Temporary Databases

### CODE REVIEW

**Comment by Lirikl** [comment]
TemporaryDatabases are mainly working, TEMPORARY keyword allow creating databases that will be deleted upon ending session and could not be seen outside of it. TemporaryDatabases all use Atomic engine and therefore could contain tables of all engines that it could

File: src/Interpreters/InterpreterCreateQuery.cpp [hunk 1]
```diff
@@ -113,6 +113,30 @@ InterpreterCreateQuery::InterpreterCreateQuery(const ASTPtr & query_ptr_, Contex
 BlockIO InterpreterCreateQuery::createDatabase(ASTCreateQuery & create)
 {
     String database_name = create.getDatabase();
+    if (create.temporary && create.storage && create.storage->engine && create.storage->engine->name != "Atomic")
+        throw Exception("Temporary database must have Atomic engine", ErrorCodes::INCORRECT_QUERY); ///Probably needs new error code
```
Comment (tavplubix) [inline src/Interpreters/InterpreterCreateQuery.cpp:117]: `INCORRECT_QUERY` is fine (`NOT_IMPLEMENTED` is suitable too)

File: src/Interpreters/InterpreterCreateQuery.cpp [hunk 2]
```diff
@@ -113,6 +113,30 @@ InterpreterCreateQuery::InterpreterCreateQuery(const ASTPtr & query_ptr_, Contex
...
+
+    String temporary_database_name = "";
+    TemporaryDatabaseHolder temporary_database_holder(getContext(), "", "");
+    if (create.temporary)
+    {
+        auto guard = DatabaseCatalog::instance().getDDLGuard(database_name, "");
```
Comment (tavplubix) [inline src/Interpreters/InterpreterCreateQuery.cpp:123]: Notice that it also locks non-temporary database with this name

File: src/Interpreters/InterpreterCreateQuery.cpp [hunk 3]
```diff
@@ -113,6 +113,30 @@ InterpreterCreateQuery::InterpreterCreateQuery(const ASTPtr & query_ptr_, Contex
...
+                return {};
+            else
+                throw Exception("Database " + database_name + " already exists.", ErrorCodes::DATABASE_ALREADY_EXISTS);
+        }
+        temporary_database_name = database_name;
+        auto id = UUIDHelpers::generateV4();
+        database_name = DatabaseCatalog::TEMPORARY_DATABASES_PREFIX + toString(id);
+        create.setDatabase(database_name);
+        temporary_database_holder = TemporaryDatabaseHolder(getContext(), temporary_database_name, database_name);
+        //create.uuid = UUIDHelpers::generateV4();
```
Comment (tavplubix) [inline src/Interpreters/InterpreterCreateQuery.cpp:138]: Why not?

File: src/Interpreters/InterpreterCreateQuery.cpp [hunk 4]
```diff
@@ -113,6 +113,30 @@ InterpreterCreateQuery::InterpreterCreateQuery(const ASTPtr & query_ptr_, Contex
...
+    }
 
     auto guard = DatabaseCatalog::instance().getDDLGuard(database_name, "");
```
Comment (tavplubix) [inline src/Interpreters/InterpreterCreateQuery.cpp:141]: Seems like a race condition is possible

File: src/Interpreters/InterpreterDropQuery.cpp [hunk 5]
```diff
@@ -294,14 +294,24 @@ BlockIO InterpreterDropQuery::executeToTemporaryTable(const String & table_name,
...
     {
-        res = executeToDatabaseImpl(query, database, tables_to_wait);
+        String db_name = getContext()->resolveTemporaryDatabase(query.getDatabase());
+        if (db_name == query.getDatabase())
+        {
+            res = executeToDatabaseImpl(query, database, tables_to_wait);
+        }
+        else
+        {
+            res = executeToTemporaryDatabase(query);
```
Comment (tavplubix) [inline src/Interpreters/InterpreterDropQuery.cpp:311]: Race condition is possible (executeToDatabaseImpl may be called for temporary database and vice versa)

File: src/Interpreters/InterpreterDropQuery.cpp [hunk 6]
```diff
@@ -294,14 +294,24 @@ BlockIO InterpreterDropQuery::executeToTemporaryTable(const String & table_name,
...
+            query.setDatabase(db_name);
```
Comment (tavplubix) [inline src/Interpreters/InterpreterDropQuery.cpp:312]: What is the purpose of setting the database name after executing the query?

File: src/Interpreters/TemporaryDatabaseHolder.cpp [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+            TableExclusiveLockHolder table_lock;
+            if (database->getUUID() == UUIDHelpers::Nil)
+                table_lock = table->lockExclusively(context_->getCurrentQueryId(), context_->getSettingsRef().lock_acquire_timeout);
+            DatabaseCatalog::instance().tryRemoveLoadingDependencies(table_id, context_->getSettingsRef().check_table_dependencies,
+                                                                     is_drop_or_detach_database);
+            database->dropTable(context_, table_id.table_name, query.sync);
+        }
+
+        db = database;
+        uuid_to_wait = table_id.uuid;
```
Comment (tavplubix) [inline src/Interpreters/TemporaryDatabaseHolder.cpp:81]: Copy-paste from `InterpreterDropQuery::executeToTableImpl`

File: src/Interpreters/TemporaryDatabaseHolder.cpp [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+                executeToTableImpl(table_context, query_for_table, db, table_to_wait);
+            }
+        }
+
+        /// Protects from concurrent CREATE TABLE queries
+        auto db_guard = DatabaseCatalog::instance().getExclusiveDDLGuardForDatabase(database_name);
+
+
+        /// DETACH or DROP database itself
+        DatabaseCatalog::instance().detachDatabase(getContext(), database_name, true /*drop*/, database->shouldBeEmptyOnDetach());
```
Comment (tavplubix) [inline src/Interpreters/TemporaryDatabaseHolder.cpp:152]: Copy-paste from `InterpreterDropQuery::executeToDatabaseImpl`

**Review by tavplubix — CHANGES_REQUESTED** [review]
> TemporaryDatabases are mainly working

Tests are required to make sure that it actually works.

> will be deleted upon ending session

What happens if server is killed or power turned off before session destructor is called?

> TemporaryDatabases all use Atomic engine and therefore could contain tables of all engines that it could

This is questionable, because some table engines are complicated and were not supposed to be temporary. In particular, they may leave garbage on local filesystem and in ZooKeeper, so proper cleanup is required. 

Maybe we should start from the first step described in https://github.com/ClickHouse/ClickHouse/issues/31497: support more engines for temporary tables

And let's avoid copy-paste where possible

**Pull request closed without merging**